runtime/src/message_processor.rs

```diff
@@ -297,26 +307,74 @@ impl<'a> ThisInvokeContext<'a> {
             timings: ExecuteDetailsTimings::default(),
-        }
+        };
+        invoke_context
+            .invoke_stack
+            .push(InvokeContextStackFrame::new(*program_id, &keyed_accounts));
+        invoke_context
     }
 }
 impl<'a> InvokeContext for ThisInvokeContext<'a> {
-    fn push(&mut self, key: &Pubkey) -> Result<(), InstructionError> {
-        if self.program_ids.len() > self.bpf_compute_budget.max_invoke_depth {
+    fn push(
+        &mut self,
+        key: &Pubkey,
+        keyed_accounts: &[(&Pubkey, bool, bool, &RefCell<AccountSharedData>)],
+    ) -> Result<(), InstructionError> {
+        if self.invoke_stack.len() > self.bpf_compute_budget.max_invoke_depth {
             return Err(InstructionError::CallDepth);
         }
-        if self.program_ids.contains(key) && self.program_ids.last() != Some(key) {
+        let frame_index = self.invoke_stack.iter().position(|frame| frame.key == *key);
+        if frame_index != None && frame_index != Some(self.invoke_stack.len().saturating_sub(1)) {
             // Reentrancy not allowed unless caller is calling itself
             return Err(InstructionError::ReentrancyNotAllowed);
         }
-        self.program_ids.push(*key);
+        // Alias the keys and account references in the provided keyed_accounts
+        // with the ones already existing in self, so that the lifetime 'a matches.
+        let keyed_accounts = keyed_accounts
+            .iter()
+            .map(|(search_key, is_signer, is_writable, account)| {
+                self.account_deps
+                    .iter()
+                    .map(|(key, _account)| key)
+                    .chain(self.message.account_keys.iter())
+                    .position(|key| key == *search_key)
+                    .map(|mut index| {
+                        if index < self.account_deps.len() {
+                            (
+                                &self.account_deps[index].0,
+                                *is_signer,
+                                *is_writable,
+                                &self.account_deps[index].1 as &RefCell<AccountSharedData>,
+                            )
+                        } else {
+                            index = index.saturating_sub(self.account_deps.len());
+                            (
+                                &self.message.account_keys[index],
+                                *is_signer,
+                                *is_writable,
+                                // TODO
+                                // Currently we are constructing new accounts on the stack
+                                // before calling MessageProcessor::process_cross_program_instruction
+                                // Ideally we would recycle the existing accounts here like this:
+                                // &self.accounts[index] as &RefCell<AccountSharedData>,
+                                unsafe { std::mem::transmute(*account) },
```

Comment:
Lets get rid of this last unsafe.  If there are other changes that have to go in first, then lets do those.

Reply:
Well, as discussed in the header of this PR, this one is really tricky. Unless I am missing something, there is no way for the `PreAccount`s `verify` to distinguish between "unlawful" changes by one of the programs and the seeming "changes" caused by correctly aliasing the accounts. The one solution I can see is to remove the need for `PreAccount`s, but that depends on this PR being done and the planned `KeyedAccounts` mapping too.

Reply:
@Lichtso would it help the lifetime issue if we reworked KeyedAccount to remove the pubkey ref and put Account behind an Rc, thus removing the explicitly 'a lifetime?